### ISSUE TITLE
Quasi Newton methods failing with GPU calculations

### ISSUE THREAD

**Issue by luk-f-a** [itdxer/neupy#108]
Hi,
I've been training my network  with Quasi Newton algorithm (bfgs mostly) on my cpu and it works really well. When I switched Theano to run on my GPU and both bfgs and sr1 (haven't tried the other ones) return NaN after the first epoch.
I tracked the NaN to 
self.predict(input_test) = 
array([[ nan],
       [ nan],
       [ nan],
       ..., 
       [ nan],
       [ nan],
       [ nan]], dtype=float32)

What I see on the console output is the first train loss to be ok (a number of the magnitude I expect) but with a NaN (in the console a "-") as test loss. Every epoch afterwards has also "-" as both loss and error loss. I am happy to run some tests to help pinpoint the problem but I don't know how to debug within a Theano function, so I haven't been able to track the NaN backwards from the output layer towards the inputs.

Let me know what more information I can provide to help with this bug.

**Comment by itdxer** [comment]
Hi @luk-f-a,

So as far as I understood you trained everything on CPU and it worked fine. After that you, without changes in the code, switched to the GPU and run it again. In that case you've gotten these NaN. Is that correct?

Some information would be useful to have:

1. NeuPy version (``import neupy; print(neupy.__version__)``)
2. Python version
3. Script that trains the network

In case if you can share your code it would be good to have:

4. Networks structure
5. Training parameters (like ``update_function`` for the ``algorithms.QuasiNewton``) 

It would be also useful to have a data, but in case if it's impossible it would be useful to know input values distributions (min and max at least).

**Comment by itdxer** [comment]
And it's not OK that NeuPy replaces NaN with a ``-`` symbol. I'll create a separate bug for it.

**Comment by luk-f-a** [comment]
Hi @itdxer, thanks for looking into it. My script is very long because of heavy pre processing, I'll put together a jupyter notebook with the essential parts and part of the data (already processed). I'm out of town today so it might ready on Monday or Tuesday.  Thanks for the great library btw.

**Comment by itdxer** [comment]
Great, I'll be waiting for it.

**Comment by luk-f-a** [comment]
Hi @itdxer, I have the example, the notebook and the data.
Neupy version is '0.3.0.b1', theano version is '0.8.2', cuda version is 7.0. 
I'm able to do Conjugate Gradient in GPU and QuasiNewton in CPU so it seems to be the particular combination of QuasiNewton and GPU. The data for CG with GPU was exactly this same data.
This is the output I get:
Main information

[ALGORITHM] QuasiNewton

[OPTION] verbose = True
[OPTION] epoch_end_signal = None
[OPTION] show_epoch = 1
[OPTION] shuffle_data = False
[OPTION] train_end_signal = None
[OPTION] error = mse
[OPTION] addons = None
[OPTION] gradient_tol = 1e-05
[OPTION] h0_scale = 1
[OPTION] update_function = bfgs

[THEANO] Initializing Theano variables and functions.
[THEANO] Initialization finished sucessfully. It took 13.36 seconds

Start training

[TRAIN DATA] 9750 samples, feature shape: (84,)
[TRAINING] Total epochs: 5

'------------------------------------------------
| Epoch  | Train err | Valid err | Time       |
'------------------------------------------------
| 1       | 4.731240  | -         | 0.2 sec    |
| 2       | -         | -         | 0.4 sec    |
| 3       | -         | -         | 0.4 sec    |
| 4       | -         | -         | 0.5 sec    |
| 5       | -         | -         | 0.4 sec    |



[Test QuasiNewton GPU.ipynb.zip](https://github.com/itdxer/neupy/files/380502/Test.QuasiNewton.GPU.ipynb.zip)

[data.zip](https://github.com/itdxer/neupy/files/380498/data.zip)

**Comment by itdxer** [comment]
Thank you. I'll check it soon.

**Comment by luk-f-a** [comment]
I'm using Python 2.7 btw

**Comment by itdxer** [comment]
I've just noticed that your Training error is too big for the first epoch. Quasi-Newton works with Wolfe Search and I think that the main problem might be there. Would you be able to disable Wolfe Search and test everything without it?

To disable wolfe search you need to comment this line and set up a constant learning step [here](https://github.com/itdxer/neupy/blob/master/neupy/algorithms/gd/quasi_newton.py#L200).

```python
# Comment this line
# step = asfloat(line_search(phi, derphi))

# Add this
step = asfloat(0.1)
```

**Comment by luk-f-a** [comment]
These are the results (the empty validation error is ok because I'm not supplying val data). it's running now, but the error is all over, I guess it's bouncing around with the fixed step.

GPU run 1:
'------------------------------------------------
| Epoch # | Train err | Valid err | Time       |
'------------------------------------------------
| 1       | 6.135746  | -         | 11.0 ms    |
| 2       | 26.761364 | -         | 0.2 sec    |
| 3       | 88.482208 | -         | 0.2 sec    |
| 4       | 63.794647 | -         | 0.2 sec    |
| 5       | 79.734856 | -         | 0.3 sec    |
'------------------------------------------------

GPU run 2:
'------------------------------------------------
| Epoch # | Train err | Valid err | Time       |
'------------------------------------------------
| 1       | 0.320606  | -         | 10.0 ms    |
| 2       | 0.946946  | -         | 0.2 sec    |
| 3       | 0.814122  | -         | 0.2 sec    |
| 4       | 0.505105  | -         | 0.2 sec    |
| 5       | ~158      | -         | 0.2 sec    |
'------------------------------------------------

GPU run 3:
'------------------------------------------------
| Epoch # | Train err | Valid err | Time       |
'------------------------------------------------
| 1       | 0.625705  | -         | 10.0 ms    |
| 2       | 2.940904  | -         | 0.2 sec    |
| 3       | 2.289172  | -         | 0.2 sec    |
| 4       | ~220      | -         | 0.3 sec    |
| 5       | 5.578112  | -         | 0.2 sec    |
'------------------------------------------------
CPU run:

'------------------------------------------------
| Epoch # | Train err | Valid err | Time       |
'------------------------------------------------
| 1       | 0.627899  | -         | 26.0 ms    |
| 2       | 8.591180  | -         | 0.9 sec    |
| 3       | 6.044182  | -         | 0.9 sec    |
| 4       | 4.975187  | -         | 0.9 sec    |
| 5       | ~148      | -         | 0.9 sec    |
'------------------------------------------------

**Comment by itdxer** [comment]
Probably step size is too big, you might try to run it with 0.01 or even less. Nevertheless, it looks like that problem with Wolfe Search. Thank you. I'll continue investigating this issue.

**Comment by luk-f-a** [comment]
I did try with smaller sizes but it didn't seem to make any difference, it would oscillate. I tried 0.01 and 0.001, I'll try something smaller.

**Comment by itdxer** [comment]
Hi @luk-f-a,

Sorry for the delay, I've been really busy for a last week. I finally tested everything on the GPU and I got the same issues. I fixed this problem on my side, but you need to try if it works for you.

Fix is located in the `release/v0.3.1` branch. To install NeuPy from this branch you need to run the following command

```bash
$ pip install git+https://github.com/itdxer/neupy.git@release/v0.3.1
```

**Comment by luk-f-a** [comment]
works great, thanks @itdxer !!!!

**Comment by itdxer** [comment]
Fixed in release/v0.3.1
